sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java

```diff
@@ -0,0 +1,193 @@
+    PAssert.that(output).satisfies(new SerializableFunction<Iterable<byte[]>, Void>() {
+
+      int count = 0;
+
+      @Override
+      public Void apply(Iterable<byte[]> input) {
+        for (byte[] element : input) {
+          String inputString = new String(element);
+          Assert.assertEquals("This is test " + count, inputString);
+          count++;
+        }
+        return null;
+      }
+    });
+
+    // produce messages on the broker in another thread
+    // This thread prevents to block the pipeline waiting for new messages
+    Thread thread = new Thread() {
+      public void run() {
+        try {
+          // gives time to the pipeline to start
+          Thread.sleep(2000);
+        } catch (Exception e) {
+          // nothing to do
+        }
+        try {
+          MqttClient client = new MqttClient("tcp://localhost:11883", "publisher");
+          client.connect();
+          for (int i = 0; i < 10; i++) {
+            MqttMessage message = new MqttMessage();
+            message.setQos(0);
+            message.setRetained(false);
+            message.setPayload(("This is test " + i).getBytes());
+            client.publish("READ_TOPIC", message);
+          }
+          client.disconnect();
+          client.close();
+        } catch (Exception e) {
+          // nothing to do
+        }
+      }
+    };
+    thread.start();
+
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWrite() throws Exception {
+    List<MqttMessage> messages = new ArrayList<>();
+    MqttClient client = receive(messages);
+
+    Pipeline pipeline = TestPipeline.create();
+
+    ArrayList<byte[]> data = new ArrayList<>();
+    for (int i = 0; i < 100; i++) {
+      data.add("Test".getBytes());
+    }
+    // we use QoS 2 here to be sure the subscriber completely receive all messages before
+    // shutting down the MQTT broker.
+    // Quality of Service 2 indicates that a message should be delivered once. The message will
+    // be persisted to disk, and will be subject to a two-phase ack.
+    pipeline.apply(Create.of(data))
+        .apply(MqttIO.write()
+            .withMqttConnectionConfiguration(
+                MqttIO.MqttConnectionConfiguration.create(
+                    "tcp://localhost:11883",
+                    "BEAM_PIPELINE",
+                    "WRITE_TOPIC"))
+            .withQoS(2));
+    pipeline.run();
+
+    Assert.assertEquals(100, messages.size());
+
+    client.disconnect();
+    client.close();
+  }
+
+  private MqttClient receive(final List<MqttMessage> messages) throws MqttException {
+    MqttClient client = new MqttClient("tcp://localhost:11883", "receiver");
+    MqttCallback callback = new MqttCallback() {
+      @Override
+      public void connectionLost(Throwable cause) {
+        cause.printStackTrace();
+      }
+
+      @Override
+      public void messageArrived(String topic, MqttMessage message) throws Exception {
+        synchronized (messages) {
```

Comment:
Nobody else is synchronizing on "messages", so this method is synchronizing against itself, which is pointless because it is invoked sequentially anyway. Which thread safety guarantee is this synchronized block trying to accomplish?